text/0000-dockerfiles.md

```diff
@@ -0,0 +1,170 @@
+2. The only valid format is `Dockerfile`, although support for, e.g. LLB JSON, could be added in the future.
+3. If the `.out` suffix is present and the file is executable, the hook is executed in the context of the app directory, and its output must match `<format>`.
+4. If the `.out` suffix is not present, the contents of the file must match `<format>`.
+
+Hook files are evaluated (either read or executed) and the resulting Dockerfiles are applied to the build-time and runtime base images in lexographical order by `<name>`.
+Executable hook files must not modify the app directory when executed and may be executed in parallel.
+However, the resulting Dockerfiles must not be applied in parallel.
+A Dockerfile intended for the runtime base image is applied after the detection phase.
+A Dockerfile intended for the build-time base image is applied before the detection phase.
+
+If an executable hook exits with a non-zero status value, the build fails.
+If a executable hook exits with a zero status value and no output, the hook is ignored. 
+Directories are ignored.
+Files at the top-level of `/cnb/hooks.d/` that do not match the specified file name format result in a build failure.
+
+Both Dockerfiles must accept `base_image` and `build_id` args.
+The `base_image` arg allows the lifecycle to specify the original base image.
+The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`. When the `$build_id` arg is referenced in a `RUN` instruction, all subsequent layerrs will be rebuilt on the next build (as the value will change).
+
+A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
+
+
+#### Example: App-specified Dockerfile Hook
+
+This example hook would allow an app to provide runtime and build-time base image extensions as "run.Dockerfile" and "build.Dockerfile."
+The app developer can decide whether the extensions are rebasable.
+
+##### `/cnb/hooks.d/app.build.Dockerfile.out`
+```
+#!/bin/sh
+cat build.Dockerfile
+```
+##### `/cnb/hooks.d/app.run.Dockerfile.out`
+```
+#!/bin/sh
+cat run.Dockerfile
+```
+
+#### Example: RPM Dockerfile Hook
+
+This example hook would allow a builder to install RPMs for each language runtime.
+
+Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.
+
+##### `/cnb/hooks.d/app.Dockerfile.out`
+```
+#!/bin/sh
+[[ -f Gemfile.lock ]] && cat /cnb/hooks.d/app.Dockerfile.d/Dockerfile-ruby
+[[ -f package.json ]] && cat /cnb/hooks.d/app.Dockerfile.d/Dockerfile-node
+```
+
+### Platform-specified Dockerfiles
```

Comment:
Will the platform-specified Dockerfiles be defined part of the builder.toml ?

Reply:
No, the section probably needs a better name.

These Dockerfiles would be inputs to `pack create-stack` (or whatever we call the helper command that will be used to create buildpack-compatible base images).

Reply:
> These Dockerfiles would be inputs to `pack create-stack` (or whatever we call the helper command that will be used to create buildpack-compatible base images).

Ideally they should be defined part of the builder.toml file as URI or passed as command parameter(s) pointing to the Dockerfiles

Reply:
> Ideally they should be defined part of the builder.toml file as URI or passed as command parameter(s) pointing to the Dockerfiles
I think you're confusing this section with the section above. A better title for this section would be something like "Using Dockerfiles to Extend Base Images." The point this that the same Dockerfiles used for extending builds could also be used to create/extend base images.